README.md
# Practical_and_Accurate
A clone of ZHAireS modified to allow for the Practical and Accurate approach to coherent radio calculations

# Methodology
The Practical and Accurate (P&A) approach to calculating radio emission from air showers involves the summation of particle tracks within 4-D volumes in spacetime with sides that adhere to the Fraunhofer limit. A basic description of the physics involved is detailed in the attached proceedings from ICRC2023.

The bulk of the methodology is found in /src/aires/Cpp_GridMaker_2.cpp. Tracks are generated in src/aires/eplusmins.f and passed to the C++ script using the "update_vector" function (using iso c binding), where they are histogrammed into non-uniform 4D bins in spacetime. In src/aires/Aires.f, after all the generated tracks have been passed, the function "calculate_radio" is called, which calculates the electric fields from the summed currents. The final text file containing the calculated electric fields is saved in the directory where the shower is generated and can be compared directly to the "timefresnel-root.dat" file if radio calculations are toggled on during runtime.

# Prerequisites:
* For running ZHAireS: I don't think any? If I'm wrong on this, update this section!  
* For plotting: numpy, matplotlib, scipy  
* For easy steering (recommended): [Remy Prechelt's zhaires.py python wrapper for ZHAireS](https://github.com/rprechelt/zhaires.py)

# Building P&A ZHAireS:
1) Compile the src/aires/grid.f90 linking script: "gfortran -c grid.f90"
2) Compile the src/aires/Cpp_gridmaker_2.cpp P&A C++ code: "g++ -c Cpp_GridMaker_2.cpp"
3) Perform the ZHAireS install: "./doinstall 0" or follow the ZHAireS "Install.HowTo" document

# How to use P&A ZHAireS:
Exactly the same as ZHAireS (for now). Either use a steering file or the zhaires.py python wrapper. An example using the latter is given in "example_runfile.py". Should you use this runfile, ensure that the absolute pathing to the Aires installation is correct for your system.


HISTORY


                    HISTORY OF AIRES SIMULATION SYSTEM.

***

12/Dec/2006 AIRES version 2.8.4a (about 730 routines and more than 101,000
----------- lines of source code).
            Functionally equivalent to AIRES 2.8.4. A minor modification in
the installation procedure with gfortran compiler is included, together with
technical bug fixes.

09/Oct/2006 AIRES version 2.8.4 (about 730 routines and more than 101,000
----------- lines of source code).
            Some differences with the previous version:
            1) Hadronic model QGSJETII-3
            2) Several bug fixes.
In all other aspects AIRES 2.8.4 is similar to AIRES 2.8.2a.

13/Dec/2005 AIRES version 2.8.2a (about 730 routines and more than 101,000
----------- lines of source code).
            A bug in the AIRES-QGSJET-II link routines has been fixed
(Affecting only showers initiated by primary nuclei).
In all other aspects AIRES 2.8.2a is similar to AIRES 2.8.2.

01/Nov/2005 AIRES version 2.8.2 (about 730 routines and more than 101,000
----------- lines of source code).
            Some differences with the previous version:
   (1) A link for QGSJET-II (beta version) is provided. A special executable
   "AiresQIIbeta" is generated during the installation process.
   (2) The default QGSJET model (executable "AiresQ") is QGSJET01c.
   (3) Geomagnetic field is calculated using the new IGRF10 model.
   (4) A few new additions to the Aires library (pending documentation!).
   (5) As always, numerous minor bugs have been fixed.

23/Mar/2005 AIRES version 2.8.0 (about 725 routines and 101,000 lines of
----------- source code).
            Some differences with the previous version:
   (1) SIBYLL 1.6 and QGSJET 99 are no longer included as hadronic model
   options.
   (2) A "no thinning" option for compiling the simulations programs is
   included.
   (3) A new program, AiresMerge, to merge the data contained in more than
   one IDF or ADF file is included.
   (4) Added numerous IDL instructions and library routines. Documentation
   of new features is missing.
   (5) Support for dynamically added fields to compressed file records.
   (6) As always, numerous minor bugs have been fixed.

11/Jul/2002 AIRES version 2.6.0 (about 670 routines and 94,000 lines of
----------- source code).
            Differences with the previous version:
   (1) New version of SIBYLL (SIBYLL 2.1) hadronic model.
   (2) New version of QGSJET (QGSJET01) hadronic model.
   (3) Nucleus-nucleus cross sections and nuclear fragmentation are now
   processed via the external hadronic packages (SIBYLL or QGSJET).
   (4) The algorithm saving ground particles into compressed files were
   improved to exactly account the arrival times of heavy neutral particles
   like neutrons.
   (5) Nuclear codes were extended beyond iron, up to Z = 36.
   (6) The simulation and summary programs can generate a task summary script
   file (extension .tss) containing the most relevant data associated with the
   simulations, in the format Keyword = value, suitable for processing by other
   programs.
   (7) Complete revision of the AIRES Runner System, focused on improving the
   capacity of the System to cope with large scale computations.
   (8) The AIRES runner scripts now support the inclusion of BeforeProcess
   macro.
   (9) New modules added to the AIRES object library, in particular some
   routines for easier processing of longitudinal particle tracking files.
   (10) Revision of the user's manual and inclusion of several issues
   that were introduced at the previous release of AIRES with pending
   documentation.
   (11) Additionally, several minor changes, improvements, and -of course-
   corrections of bugs!

19/Oct/2001 AIRES version 2.4.0 (about 620 routines and 88,000 lines of
----------- source code).
            Differences with the previous version:
   (1) New version of the splitting algorithm used to process low energy
   hadronic interactions. This new version was tuned to effectively
   emulate well-known interaction models.
   (2) Low energy cross sections are now evaluated using fits to experimental
   data.
   (3) Improved procedures for gamma-nucleus collisions.
   (4) Muon propagation now takes into account muon bremsstrahlung and
   muonic pair production. The new algorithms are based in the theory by
   Kokoulin and Petrukhin.
   (5) New improved procedure to treat knock-on electrons generated by heavy
   charged particles.
   (6) Full propagation of Lambda baryons.
   (7) Improved procedures for meson and baryon decay.
   (8) Exhaustive revision of the electromagnetic shower engine.
   (9) The thinning algorithm posseses now two maximum weight parameters,
   respectively for electromagnetic (gamma, e+, e-) and heavy particles.
   (10) New IGRF 2000 data set used to estimate geomagnetic field.
   (11) 48 new longitudinal development tables corresponding to the following
   observables: Number of low energy particles versus atmospheric depth;
   Energy of low energy particles versus atmospheric depth; deposited energy
   (ionization) versus atmospheric depth.
   (12) Improved procedure for updating idf and compressed files, that
   permits more stable operation during very long runs.
   (13) Many new modules in the AIRES library.
   (14) Global variables can be defined within the input data. Such variables
   can be used for text replacement within the input file and/or be passed
   to external analysis or special primary modules.
   (15) New length (inches, feet, yards, and miles) and energy (Joules) units
   supported. The standard "keV" replaces the old "KeV" which is maintained
   as valid for backwards compatibility.
   (16) Additionally, lots of minor changes, improvements, and -of course-
   corrections of bugs!

23/Dec/1999 AIRES version 2.2.1. Functionally equivalent to version 2.2.0.
----------- A bug in the QGSJET interface (Diffractive interactions were
            mistakenly disabled) has been fixed.


src/misc/psi.f
c
c     FILE: psi.f                         Creation date: 26/OCT/2004.
c                                       LAST MODIFICATION: 26/OCT/2004.
c
c     A routine to evaluate the Psi function (derivative of the
c     logarithm of the Gamma function) for real x
c
c<---><---><---><---><---><---><--->*<---><---><---><---><---><---><--->
c<---><---><---><---><---><---><--->*<---><---><---><---><---><---><--->
c
      function Psi (x)
c
c     Derivative of the logarithm of the Gamma function for real x.
c
c     S. J. Sciutto, La Plata 1994, 2004.
c
      implicit none
c
      double precision  Psi
      double precision  x
c
      integer           m, i
      double precision  y, prec, ssum
c
c     Values of Riemman's zeta function
c
      double precision  rzeta(2:55)
c
      data rzeta(02) /  1.64493406684822643647d0 / 
      data rzeta(03) / -1.20205690315959428540d0 /
      data rzeta(04) /  1.08232323371113819152d0 /
      data rzeta(05) / -1.03692775514336992633d0 /
      data rzeta(06) /  1.01734306198444913971d0 /
      data rzeta(07) / -1.00834927738192282684d0 /
      data rzeta(08) /  1.00407735619794433938d0 /
      data rzeta(09) / -1.00200839282608221442d0 /
      data rzeta(10) /  1.00099457512781808534d0 /
      data rzeta(11) / -1.00049418860411946456d0 /
      data rzeta(12) /  1.00024608655330804830d0 /
      data rzeta(13) / -1.00012271334757848915d0 /
      data rzeta(14) /  1.00006124813505870483d0 /
      data rzeta(15) / -1.00003058823630702049d0 /
      data rzeta(16) /  1.00001528225940865187d0 /
      data rzeta(17) / -1.00000763719763789976d0 /
      data rzeta(18) /  1.00000381729326499984d0 /
      data rzeta(19) / -1.00000190821271655394d0 /
      data rzeta(20) /  1.00000095396203387280d0 /
      data rzeta(21) / -1.00000047693298678781d0 /
      data rzeta(22) /  1.00000023845050272773d0 /
      data rzeta(23) / -1.00000011921992596531d0 /
      data rzeta(24) /  1.00000005960818905126d0 /
      data rzeta(25) / -1.00000002980350351465d0 /
      data rzeta(26) /  1.00000001490155482837d0 /
      data rzeta(27) / -1.00000000745071178984d0 /
      data rzeta(28) /  1.00000000372533402479d0 /
      data rzeta(29) / -1.00000000186265972351d0 /
      data rzeta(30) /  1.00000000093132743242d0 /
      data rzeta(31) / -1.00000000046566290650d0 /
      data rzeta(32) /  1.00000000023283118337d0 /
      data rzeta(33) / -1.00000000011641550173d0 /
      data rzeta(34) /  1.00000000005820772088d0 /
      data rzeta(35) / -1.00000000002910385044d0 /
      data rzeta(36) /  1.00000000001455192189d0 /
      data rzeta(37) / -1.00000000000727595984d0 /
      data rzeta(38) /  1.00000000000363797955d0 /
      data rzeta(39) / -1.00000000000181898965d0 /
      data rzeta(40) /  1.00000000000090949478d0 /
      data rzeta(41) / -1.00000000000045474738d0 /
      data rzeta(42) /  1.00000000000022737368d0 /
      data rzeta(43) / -1.00000000000011368684d0 /
      data rzeta(44) /  1.00000000000005684342d0 /
      data rzeta(45) / -1.00000000000002842171d0 /
      data rzeta(46) /  1.00000000000001421085d0 /
      data rzeta(47) / -1.00000000000000710543d0 /
      data rzeta(48) /  1.00000000000000355271d0 /
      data rzeta(49) / -1.00000000000000177636d0 /
      data rzeta(50) /  1.00000000000000088818d0 /
      data rzeta(51) / -1.00000000000000044409d0 /
      data rzeta(52) /  1.00000000000000022204d0 /
      data rzeta(53) / -1.00000000000000011102d0 /
      data rzeta(54) /  1.00000000000000005551d0 /
      data rzeta(55) / -1.00000000000000002776d0 /
c
c     Euler's constant
c
      double precision  Eulerneg
      data              Eulerneg  / -0.577215664901532860607d0 /
c
c     FIRST EXECUTABLE STATEMENT.
c
      y = abs(x - 0.5d0)
      m = y
      if (x .ge. 0.5d0) then
        prec = 0
        do i = 1, m
          prec = prec + 1.d0 / (x - i)
        enddo
        y = x - (m + 1)
      else
        if (m .eq. y) m = m - 1
        prec = -1.d0 / x
        do i = 1, m
          prec = prec - 1.d0 / (x + i)
        enddo
        y = x + m
      endif
c
c     Evaluating the series.
c
      ssum = rzeta(55)
      do i = 54, 2, -1
        ssum = rzeta(i) + y * ssum
      enddo
c
      Psi = prec + y * ssum + Eulerneg
      return
      end
c     --- End of routine Gamma
c
c<---><---><---><---><---><---><--->*<---><---><---><---><---><---><--->
c<---><---><---><---><---><---><--->*<---><---><---><---><---><---><--->
c
c     End of file 'psi.f'
c     This source file is part of AIRES 2.8.4a distribution.
c
c     Ay49299427284ay 0 Tue Dec 12 16:29:49 ART 2006
